File: app/java/app/src/main/java/com/example/accounting/application/service/FinancialStatementService.java
package com.example.accounting.application.service;

import com.example.accounting.application.port.out.FinancialDataRepository;
import com.example.accounting.application.model.financial.BalanceSheet;
import com.example.accounting.application.model.financial.BalanceSheetItem;
import com.example.accounting.application.model.financial.FinancialRatios;
import com.example.accounting.application.model.financial.IncomeStatement;
import com.example.accounting.application.model.financial.IncomeStatementItem;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.List;

/**
 * 財務諸表生成サービス（Application Service）
 */
@Service
@Transactional(readOnly = true)
public class FinancialStatementService {

    private final FinancialDataRepository financialDataRepository;

    public FinancialStatementService(FinancialDataRepository financialDataRepository) {
        this.financialDataRepository = financialDataRepository;
    }

    /**
     * 貸借対照表を生成する
     *
     * @param asOfDate 基準日
     * @return 貸借対照表
     */
    public BalanceSheet generateBalanceSheet(LocalDate asOfDate) {
        // 資産を取得（取引要素区分='1'）
        List<BalanceSheetItem> assets = financialDataRepository.findBalanceSheetItems(asOfDate, "1");

        // 負債を取得（取引要素区分='2'）
        List<BalanceSheetItem> liabilities = financialDataRepository.findBalanceSheetItems(asOfDate, "2");

        // 純資産を取得（取引要素区分='3'）
        List<BalanceSheetItem> equity = financialDataRepository.findBalanceSheetItems(asOfDate, "3");

        // 合計を計算
        BigDecimal totalAssets = calculateTotal(assets);
        BigDecimal totalLiabilities = calculateTotal(liabilities);
        BigDecimal totalEquity = calculateTotal(equity);
        BigDecimal totalLiabilitiesAndEquity = totalLiabilities.add(totalEquity);

        // 構成比率を計算
        calculateRatios(assets, totalAssets);
        calculateRatios(liabilities, totalLiabilitiesAndEquity);
        calculateRatios(equity, totalLiabilitiesAndEquity);

        return new BalanceSheet(
                asOfDate,
                assets,
                liabilities,
                equity,
                totalAssets,
                totalLiabilities,
                totalEquity,
                totalLiabilitiesAndEquity
        );
    }

    /**
     * 貸借対照表の合計を計算
     */
    private BigDecimal calculateTotal(List<BalanceSheetItem> items) {
        return items.stream()
                .map(BalanceSheetItem::getBalance)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    /**
     * 構成比率を計算（%）
     */
    private void calculateRatios(List<BalanceSheetItem> items, BigDecimal total) {
        if (total.compareTo(BigDecimal.ZERO) == 0) {
            return;
        }

        for (BalanceSheetItem item : items) {
            BigDecimal ratio = item.getBalance()
                    .multiply(new BigDecimal("100"))
                    .divide(total, 2, RoundingMode.HALF_UP);
            item.setRatio(ratio);
        }
    }

    /**
     * 損益計算書を生成する
     *
     * @param fromDate 開始日
     * @param toDate 終了日
     * @return 損益計算書
     */
    public IncomeStatement generateIncomeStatement(LocalDate fromDate, LocalDate toDate) {
        // 収益を取得（取引要素区分='4'）
        List<IncomeStatementItem> revenues = financialDataRepository.findIncomeStatementItems(
                fromDate, toDate, "4");

        // 費用を取得（取引要素区分='5'）
        List<IncomeStatementItem> expenses = financialDataRepository.findIncomeStatementItems(
                fromDate, toDate, "5");

        // 合計を計算
        BigDecimal totalRevenues = calculateIncomeStatementTotal(revenues);
        BigDecimal totalExpenses = calculateIncomeStatementTotal(expenses);

        // 売上原価（勘定科目コードが51で始まる）
        BigDecimal costOfSales = expenses.stream()
                .filter(e -> e.getAccountCode().startsWith("51"))
                .map(IncomeStatementItem::getBalance)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // 販管費（勘定科目コードが6または52で始まる）
        BigDecimal operatingExpenses = expenses.stream()
                .filter(e -> e.getAccountCode().startsWith("6") || e.getAccountCode().startsWith("52"))
                .map(IncomeStatementItem::getBalance)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // 利益項目の計算
        BigDecimal grossProfit = totalRevenues.subtract(costOfSales);
        BigDecimal operatingIncome = grossProfit.subtract(operatingExpenses);
        BigDecimal netIncome = totalRevenues.subtract(totalExpenses);

        // 対売上比を計算
        calculatePercentages(revenues, totalRevenues);
        calculatePercentages(expenses, totalRevenues);

        return new IncomeStatement(
                fromDate,
                toDate,
                revenues,
                expenses,
                grossProfit,
                operatingIncome,
                netIncome,
                totalRevenues,
                totalExpenses
        );
    }

    /**
     * 損益計算書の合計を計算
     */
    private BigDecimal calculateIncomeStatementTotal(List<IncomeStatementItem> items) {
        return items.stream()
                .map(IncomeStatementItem::getBalance)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    /**
     * 対売上比を計算（%）
     */
    private void calculatePercentages(List<IncomeStatementItem> items, BigDecimal totalRevenues) {
        if (totalRevenues.compareTo(BigDecimal.ZERO) == 0) {
            return;
        }

        for (IncomeStatementItem item : items) {
            BigDecimal percentage = item.getBalance()
                    .multiply(new BigDecimal("100"))
                    .divide(totalRevenues, 2, RoundingMode.HALF_UP);
            item.setPercentage(percentage);
        }
    }

    /**
     * 財務指標を計算する
     *
     * @param balanceSheet 貸借対照表
     * @param incomeStatement 損益計算書
     * @return 財務指標
     */
    public FinancialRatios calculateFinancialRatios(
            BalanceSheet balanceSheet,
            IncomeStatement incomeStatement) {

        // 流動資産・流動負債の抽出
        BigDecimal currentAssets = balanceSheet.getAssets().stream()
                .filter(asset -> asset.getAccountCode().startsWith("11")
                        || asset.getAccountCode().startsWith("12"))
                .map(BalanceSheetItem::getBalance)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        BigDecimal currentLiabilities = balanceSheet.getLiabilities().stream()
                .filter(liability -> liability.getAccountCode().startsWith("21"))
                .map(BalanceSheetItem::getBalance)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        // 各種指標の計算
        BigDecimal currentRatio = currentLiabilities.compareTo(BigDecimal.ZERO) > 0
                ? currentAssets.divide(currentLiabilities, 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal debtToEquityRatio = balanceSheet.getTotalAssets().compareTo(BigDecimal.ZERO) > 0
                ? balanceSheet.getTotalEquity()
                        .divide(balanceSheet.getTotalAssets(), 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal grossProfitMargin = incomeStatement.getTotalRevenues().compareTo(BigDecimal.ZERO) > 0
                ? incomeStatement.getGrossProfit()
                        .divide(incomeStatement.getTotalRevenues(), 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal operatingProfitMargin = incomeStatement.getTotalRevenues().compareTo(BigDecimal.ZERO) > 0
                ? incomeStatement.getOperatingIncome()
                        .divide(incomeStatement.getTotalRevenues(), 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal netProfitMargin = incomeStatement.getTotalRevenues().compareTo(BigDecimal.ZERO) > 0
                ? incomeStatement.getNetIncome()
                        .divide(incomeStatement.getTotalRevenues(), 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal roa = balanceSheet.getTotalAssets().compareTo(BigDecimal.ZERO) > 0
                ? incomeStatement.getNetIncome()
                        .divide(balanceSheet.getTotalAssets(), 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        BigDecimal roe = balanceSheet.getTotalEquity().compareTo(BigDecimal.ZERO) > 0
                ? incomeStatement.getNetIncome()
                        .divide(balanceSheet.getTotalEquity(), 10, RoundingMode.HALF_UP)
                        .multiply(new BigDecimal("100"))
                        .setScale(2, RoundingMode.HALF_UP)
                : BigDecimal.ZERO;

        return new FinancialRatios(
                currentRatio,
                debtToEquityRatio,
                grossProfitMargin,
                operatingProfitMargin,
                netProfitMargin,
                roa,
                roe
        );
    }
}

File: app/scala/src/main/scala/com/example/accounting/application/service/FinancialStatementService.scala
package com.example.accounting.application.service

import com.example.accounting.application.port.in.FinancialStatementUseCase
import com.example.accounting.domain.financial.*
import scalikejdbc.*

import java.time.LocalDate
import scala.math.BigDecimal.RoundingMode

/**
 * 財務諸表生成サービス（FinancialStatementUseCase の実装）
 */
class FinancialStatementService extends FinancialStatementUseCase:

  override def generateBalanceSheet(asOfDate: LocalDate)(implicit session: DBSession): BalanceSheet =
    // 資産・負債・純資産の残高を取得
    val balances = sql"""
      SELECT
        a."勘定科目コード" as account_code,
        a."勘定科目名" as account_name,
        a."BSPL区分" as bspl_type,
        a."勘定科目種別"::text as account_type,
        COALESCE(SUM(d."借方金額"), 0) - COALESCE(SUM(d."貸方金額"), 0) as balance
      FROM "勘定科目マスタ" a
      LEFT JOIN "日次勘定科目残高" d
        ON a."勘定科目コード" = d."勘定科目コード"
      WHERE a."BSPL区分" = 'B'
        AND (d."起票日" <= ${asOfDate} OR d."起票日" IS NULL)
      GROUP BY a."勘定科目コード", a."勘定科目名", a."BSPL区分", a."勘定科目種別"
      HAVING COALESCE(SUM(d."借方金額"), 0) - COALESCE(SUM(d."貸方金額"), 0) != 0
      ORDER BY a."勘定科目コード"
    """
      .map(rs =>
        (
          rs.string("account_code"),
          rs.string("account_name"),
          rs.stringOpt("account_type").getOrElse(""),
          BigDecimal(rs.bigDecimal("balance")),
        )
      )
      .list
      .apply()

    // データを分類（勘定科目種別で分類）
    val (assetBalances, liabilityBalances, equityBalances) = balances.foldLeft(
      (
        List.empty[(String, String, BigDecimal)],
        List.empty[(String, String, BigDecimal)],
        List.empty[(String, String, BigDecimal)],
      )
    ) { case ((assets, liabilities, equity), (accountCode, accountName, accountType, balance)) =>
      accountType match
        case "資産" =>
          ((accountCode, accountName, balance.abs) :: assets, liabilities, equity)
        case "負債" =>
          (assets, (accountCode, accountName, balance.abs) :: liabilities, equity)
        case "純資産" | "資本" =>
          (assets, liabilities, (accountCode, accountName, balance.abs) :: equity)
        case _ =>
          (assets, liabilities, equity)
    }

    // 合計を計算
    val totalAssets = assetBalances.map(_._3).sum
    val totalLiabilities = liabilityBalances.map(_._3).sum
    val totalEquity = equityBalances.map(_._3).sum
    val totalLiabilitiesAndEquity = totalLiabilities + totalEquity

    // 構成比率を計算
    val assets = calculatePercentage(assetBalances.reverse, totalAssets)
    val liabilities = calculatePercentage(liabilityBalances.reverse, totalLiabilitiesAndEquity)
    val equity = calculatePercentage(equityBalances.reverse, totalLiabilitiesAndEquity)

    BalanceSheet(
      asOfDate = asOfDate,
      assets = assets,
      liabilities = liabilities,
      equity = equity,
      totalAssets = totalAssets,
      totalLiabilities = totalLiabilities,
      totalEquity = totalEquity,
      totalLiabilitiesAndEquity = totalLiabilitiesAndEquity,
    )

  override def generateIncomeStatement(fromDate: LocalDate, toDate: LocalDate)(implicit
      session: DBSession
  ): IncomeStatement =
    // 収益・費用の残高を取得
    val balances = sql"""
      SELECT
        a."勘定科目コード" as account_code,
        a."勘定科目名" as account_name,
        a."BSPL区分" as bspl_type,
        a."勘定科目種別"::text as account_type,
        COALESCE(SUM(d."貸方金額"), 0) - COALESCE(SUM(d."借方金額"), 0) as balance
      FROM "勘定科目マスタ" a
      LEFT JOIN "日次勘定科目残高" d
        ON a."勘定科目コード" = d."勘定科目コード"
      WHERE a."BSPL区分" = 'P'
        AND d."起票日" BETWEEN ${fromDate} AND ${toDate}
      GROUP BY a."勘定科目コード", a."勘定科目名", a."BSPL区分", a."勘定科目種別"
      HAVING COALESCE(SUM(d."貸方金額"), 0) - COALESCE(SUM(d."借方金額"), 0) != 0
      ORDER BY a."勘定科目コード"
    """
      .map(rs =>
        (
          rs.string("account_code"),
          rs.string("account_name"),
          rs.stringOpt("account_type").getOrElse(""),
          BigDecimal(rs.bigDecimal("balance")),
        )
      )
      .list
      .apply()

    // データを分類（勘定科目種別で分類）
    val (revenueBalances, expenseBalances) = balances.foldLeft(
      (List.empty[(String, String, BigDecimal)], List.empty[(String, String, BigDecimal)])
    ) { case ((revenues, expenses), (accountCode, accountName, accountType, balance)) =>
      accountType match
        case "収益" =>
          ((accountCode, accountName, balance.abs) :: revenues, expenses)
        case "費用" =>
          (revenues, (accountCode, accountName, balance.abs) :: expenses)
        case _ =>
          (revenues, expenses)
    }

    // 合計を計算
    val totalRevenues = revenueBalances.map(_._3).sum
    val totalExpenses = expenseBalances.map(_._3).sum

    // 売上原価（勘定科目コードが51で始まる）
    val costOfSales = expenseBalances
      .filter(_._1.startsWith("51"))
      .map(_._3)
      .sum

    // 販管費（勘定科目コードが6または52で始まる）
    val operatingExpenses = expenseBalances
      .filter(e => e._1.startsWith("6") || e._1.startsWith("52"))
      .map(_._3)
      .sum

    // 利益項目の計算
    val grossProfit = totalRevenues - costOfSales
    val operatingIncome = grossProfit - operatingExpenses
    val netIncome = totalRevenues - totalExpenses

    // 構成比率を計算（対売上比）
    val revenues = calculatePercentageForPL(revenueBalances.reverse, totalRevenues)
    val expenses = calculatePercentageForPL(expenseBalances.reverse, totalRevenues)

    IncomeStatement(
      fromDate = fromDate,
      toDate = toDate,
      revenues = revenues,
      expenses = expenses,
      grossProfit = grossProfit,
      operatingIncome = operatingIncome,
      netIncome = netIncome,
      totalRevenues = totalRevenues,
      totalExpenses = totalExpenses,
    )

  override def calculateFinancialRatios(
      balanceSheet: BalanceSheet,
      incomeStatement: IncomeStatement,
  ): FinancialRatios =
    // 流動資産・流動負債の抽出
    val currentAssets = balanceSheet.assets
      .filter(asset => asset.accountCode.startsWith("11") || asset.accountCode.startsWith("12"))
      .map(_.balance)
      .sum

    val currentLiabilities = balanceSheet.liabilities
      .filter(_.accountCode.startsWith("21"))
      .map(_.balance)
      .sum

    // 各種指標の計算
    val currentRatio = if currentLiabilities > 0 then
      (currentAssets / currentLiabilities * 100).setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    val equityRatio = if balanceSheet.totalAssets > 0 then
      (balanceSheet.totalEquity / balanceSheet.totalAssets * 100).setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    val grossProfitMargin = if incomeStatement.totalRevenues > 0 then
      (incomeStatement.grossProfit / incomeStatement.totalRevenues * 100)
        .setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    val operatingProfitMargin = if incomeStatement.totalRevenues > 0 then
      (incomeStatement.operatingIncome / incomeStatement.totalRevenues * 100)
        .setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    val netProfitMargin = if incomeStatement.totalRevenues > 0 then
      (incomeStatement.netIncome / incomeStatement.totalRevenues * 100)
        .setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    val roa = if balanceSheet.totalAssets > 0 then
      (incomeStatement.netIncome / balanceSheet.totalAssets * 100).setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    val roe = if balanceSheet.totalEquity > 0 then
      (incomeStatement.netIncome / balanceSheet.totalEquity * 100).setScale(2, RoundingMode.HALF_UP)
    else BigDecimal(0)

    FinancialRatios(
      currentRatio = currentRatio,
      equityRatio = equityRatio,
      grossProfitMargin = grossProfitMargin,
      operatingProfitMargin = operatingProfitMargin,
      netProfitMargin = netProfitMargin,
      roa = roa,
      roe = roe,
    )

  private def calculatePercentage(
      items: List[(String, String, BigDecimal)],
      total: BigDecimal,
  ): List[BalanceSheetItem] =
    items.map { case (accountCode, accountName, balance) =>
      val percentage =
        if total > 0 then (balance / total * 100).setScale(2, RoundingMode.HALF_UP)
        else BigDecimal(0)

      BalanceSheetItem(
        accountCode = accountCode,
        accountName = accountName,
        balance = balance,
        percentage = percentage,
      )
    }

  private def calculatePercentageForPL(
      items: List[(String, String, BigDecimal)],
      totalRevenues: BigDecimal,
  ): List[IncomeStatementItem] =
    items.map { case (accountCode, accountName, balance) =>
      val percentage =
        if totalRevenues > 0 then (balance / totalRevenues * 100).setScale(2, RoundingMode.HALF_UP)
        else BigDecimal(0)

      IncomeStatementItem(
        accountCode = accountCode,
        accountName = accountName,
        balance = balance,
        percentage = percentage,
      )
    }
